### PR TITLE
Reducing Allocation from GitInfo

### PR DESCRIPTION
**Pull request by dougqh**
# What Does This Do
Reuses TagMap.Entry objects for git repository url & commit sha

# Motivation
Reduces allocation & garbage collection - results in better throughput under high load / when memory is scarce

# Additional Notes

# Contributor Checklist

- Format the title [according the contribution guidelines](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#title-format)
- Assign the `type:` and (`comp:` or `inst:`) labels in addition to [any useful labels](https://github.com/DataDog/dd-trace-java/blob/master/CONTRIBUTING.md#labels)
- Don't use `close`, `fix` or any [linking keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) when referencing an issue.  
  Use `solves` instead, and assign the PR [milestone](https://github.com/DataDog/dd-trace-java/milestones) to the issue
- Update the [CODEOWNERS](https://github.com/DataDog/dd-trace-java/blob/master/.github/CODEOWNERS) file on source file addition, move, or deletion
- Update the [public documentation](https://docs.datadoghq.com/tracing/trace_collection/library_config/java/) in case of new configuration flag or behavior

Jira ticket: [PROJ-IDENT]

### CODE REVIEW

File: dd-java-agent/instrumentation-testing/src/main/groovy/datadog/trace/agent/test/TrackingSpanDecorator.groovy [hunk 1]
```diff
@@ -114,6 +114,11 @@ class TrackingSpanDecorator implements AgentSpan {
     return delegate.setTag(key, value)
   }
 
+  @Override
```
Comment (dougqh) [inline dd-java-agent/instrumentation-testing/src/main/groovy/datadog/trace/agent/test/TrackingSpanDecorator.groovy:117]: Most of these changes are from https://github.com/DataDog/dd-trace-java/pull/10472 which I plan to merge first

See GitInfo and GitMetadataTraceInterceptor for the "real" changes here

File: internal-api/src/main/java/datadog/trace/api/git/GitInfo.java [hunk 2]
```diff
@@ -1,21 +1,30 @@
...
+  private final TagMap.Entry repositoryEntry;
+  private final TagMap.Entry commitEntry;
+
   public GitInfo(String repositoryURL, String branch, String tag, CommitInfo commit) {
     this.repositoryURL = repositoryURL;
     this.branch = branch;
     this.tag = tag;
     this.commit = commit;
+
+    // GitInfo is reused across many traces, so create entries once and reuse them (see addTags)
```
Comment (dougqh) [inline internal-api/src/main/java/datadog/trace/api/git/GitInfo.java:25]: Despite the rest of this class being coded defensively for null - as far as I could tell from GitMetadataTraceInterceptor, repositoryURL and commit.getSha() are never null.

That said, I'm happy to add defensive here and in addTags if so desired.

File: internal-api/src/main/java/datadog/trace/api/git/GitInfo.java [hunk 3]
```diff
@@ -34,6 +43,11 @@ public CommitInfo getCommit() {
     return commit;
   }
 
+  public void addTags(AgentSpan span) {
```
Comment (dougqh) [inline internal-api/src/main/java/datadog/trace/api/git/GitInfo.java:50]: I think eventually I'd like introduce an interface for objects that add tags to span or TagMap.
Since this is the first change of this sort, I decided to share this as is to solicit thoughts on the matter.

File: internal-api/src/main/java/datadog/trace/api/git/GitInfo.java (hunk 2 above)
Comment (daniel-mohedano) [reply internal-api/src/main/java/datadog/trace/api/git/GitInfo.java:25]: In Test Optimization we've seen some edge cases in which `commit.getSha()` could be null (seems much more unlikely for repositoryURL). We usually have to point them to set the git info through environment variables in those cases. Will the `TagMap.Entry.create` call have issues with a null object?

Comment (dougqh) [reply internal-api/src/main/java/datadog/trace/api/git/GitInfo.java:25]: Good to know.

No, TagMap.Entry.create doesn't support having a null value.
In AgentSpan / DDSpan setting a null tag is translated into a removal.

TagMap does have EntryRemoval for that purpose, but EntryRemoval-s are never stored in TagMap directly just in TagMap.Ledger-s.

I think the solution is just to guard against the null-s.  I just wanted to clarify whether or not null-s were expected in the Git related code.

Comment (dougqh) [reply internal-api/src/main/java/datadog/trace/api/git/GitInfo.java:25]: In another PR, I've tweaked TagMap.Entry.create to allow null & empty String to return a null TagMap.Entry.
And now, AgentSpan / DDSpan ignore being called with a null TagMap.Entry.

So that now allows null to be handled simply in the calling code, I'll update this PR once the other one is merged.

Comment (dougqh) [reply internal-api/src/main/java/datadog/trace/api/git/GitInfo.java:25]: Done.  Thanks for the review.  You were right that commit sha can be null.
